### PR TITLE
drivers: i2c: stm32: Disable suspend to idle during the transaction

### PR DESCRIPTION
**Pull request by coran21**
Suspend-to-idle stops I2C module clocks, which must remain active during the transaction.

Waiting for semaphore inside `stm32_i2c_msg_write` and `stm32_i2c_msg_read` triggers suspend to idle when pm is enabled, which results in disabled clocks, timeout, and error.

### CODE REVIEW

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:161]: @ceolin would you mind providing some guidance here ?

I guess the case here is when only CONFIG_PM is used and hence `pm_device_busy_set()` has no effect.

Should we really implement the following in such case ?
```
#ifdef CONFIG_PM_DEVICE_RUNTIME
	(void)pm_device_runtime_get(dev);
#elif CONFIG_PM_DEVICE
	pm_device_busy_set(dev);
#else
       pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
#endif
```

**Review by erwango — CHANGES_REQUESTED** [review]
I think the proposal is correct, but highlights the complexity of PM subsystem.
I'd like PM subsystem maintainer to comment on this case.

File: drivers/i2c/i2c_ll_stm32.c
Comment (coran21) [reply drivers/i2c/i2c_ll_stm32.c:161]: I don't think that this will work because suspend to idle happens regardless of the device runtime state. We need to disable suspend to idle when power management is enabled.

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:161]: Ok, right. Point might then be that `pm_device_busy_set()` is actually mostly useless in STM32 drivers.

**Comment by coran21** [comment]
@ceolin 
Could you look at the suggested change?

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:161]: @coran21 
In Idle function `pm_system_suspend()` will be used to trigger PM related actions.
In the algo, call to `pm_policy_next_state()` will end up in `pm_policy_state_lock_is_active`.
At this step, if `pm_policy_state_lock_get()` has been previously called system suspension is cancelled, `pm_suspend_devices()` doesn't happen and `pm_device_busy_set()` has no effect in any way.

So I agree, we should keep `pm_device_runtime_get()`, but `pm_device_busy_set()` becomes a no-op according to current code and should be removed.

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:161]: @coran21 Do you have plans to continue this PR ?

Comment (coran21) [reply drivers/i2c/i2c_ll_stm32.c:161]: Yes, I would like to finish this because I2C currently doesn't work at all with power management enabled on STM32L4 (or at least for me). I was waiting for the feedback from @ceolin. Should I continue with removing `pm_device_busy_set()` as you mentioned above?

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:161]: Maybe you can have a try and confirm this is working.
On my side, I would approuve such change

Comment (coran21) [reply drivers/i2c/i2c_ll_stm32.c:161]: I can confirm that this works right away. Patching the I2C driver with these changes is the first thing I have to do to get working EEPROM in my project after I update the Zephyr source code.

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:161]: @coran21 Sorry if this wasn't clear enough, but the request was to remove `pm_device_busy_set()` and test.

Comment (coran21) [reply drivers/i2c/i2c_ll_stm32.c:155]: I removed `pm_device_busy_set()` and tested it and it's exactly how you wrote - this function doesn't have any effect if `pm_policy_state_lock_get()` was called before.

Comment (coran21) [reply drivers/i2c/i2c_ll_stm32.c:180]: It looks like there were some changes in the actual file which resulted in conflict. Could you help me how I can update my local branch `fix_stm32_i2c_disable_suspend` with these changes so that I can push the changes again?  I know that I should do some kind of rebase, but it happened to me a few times that I did a rebase and force-pushed all the previous commits, which were rebased into the pull request, and it was a mess.
In Zephyr documentation is that I should do this while I have checked out my branch with changes:
```
git fetch --all
git rebase --ignore-whitespace upstream/main
```
And do interactive rebase of my commit, is that true? I'm really not sure what to do if there is a conflict in my commit.

Comment (coran21) [reply drivers/i2c/i2c_ll_stm32.c:180]: I initially thought that I can just resolve the conflicts in the Github editor and merge them, but that didn't work.

Comment (coran21) [reply drivers/i2c/i2c_ll_stm32.c:155]: @erwango, Could you help me with that? How could I update my local branch, which has conflicts with Zephyr:main, so that it won't push all the Zephyr commits that were merged after this commit after I do force-push to this PR branch?

Comment (ceolin) [reply drivers/i2c/i2c_ll_stm32.c:155]: You are right, in this case `pm_device_busy_set` is useless if you have a lock in the power state that triggers device pm. The only thing you should keep in mind is that when a device is set busy, it won't be suspended in any power state transition. Here, the only when transitioning to `PM_STATE_SUSPEND_TO_IDLE` the device won't be suspended.

Comment (ceolin) [reply drivers/i2c/i2c_ll_stm32.c:155]: @erwango @coran21  it looks reasonable to me. If you need to lock a power state to avoid other components (clocks) to be suspended as well, setting the busy state for only this is useless. The only thing that I am thinking is if there are other power states in the target that would also trigger device power management. If there isn't any other one, than this change is correct imho.

**Comment by ceolin** [comment]
@coran21 @erwango Sorry for the late reply, I think I was not in the review list and have missed the notification. I tend to go over prs that have me listed as reviewer.

**Review by erwango — DISMISSED** [review]

**Comment by coran21** [comment]
> @coran21 Please rebase

@erwango
I want to rebase, but I'm unsure how to do that correctly because the last time I rebased the code from Zephyr:main into my local branch, it pushed all the new commits into the PR, and I had to close it. 

I wrote you a couple messages asking for help:

> It looks like there were some changes in the actual file which resulted in conflict. Could you help me how I can update my local branch fix_stm32_i2c_disable_suspend with these changes so that I can push the changes again? I know that I should do some kind of rebase, but it happened to me a few times that I did a rebase and force-pushed all the previous commits, which were rebased into the pull request, and it was a mess.
> In Zephyr documentation is that I should do this while I have checked out my branch with changes:
> 
> git fetch --all
> git rebase --ignore-whitespace upstream/main
> And do interactive rebase of my commit, is that true? I'm really not sure what to do if there is a conflict in my commit.

> @erwango, Could you help me with that? How could I update my local branch, which has conflicts with Zephyr:main, so that it won't push all the Zephyr commits that were merged after this commit after I do force-push to this PR branch?

I would really appreciate any help with this.

**Comment by erwango** [comment]
> I would really appreciate any help with this.

If there is only one commit, what you can do is:
- create a new branch temp_i2c_rebase
- check it out on top of current main
- cherry pick your commit from branch fix_stm32_i2c_disable_suspend to the new branch
- resolve the conflict
- once you're done, rename fix_stm32_i2c_disable_suspend to fix_stm32_i2c_disable_suspend_old
- rename temp_i2c_rebase to fix_stm32_i2c_disable_suspend
- force push the new stuff

**Comment by coran21** [comment]
@erwango That was exactly what I was looking for, than you!

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:180]: @gautierg-st Should we add `PM_STATE_SUSPEND_TO_RAM` here ?

Comment (gautierg-st) [reply drivers/i2c/i2c_ll_stm32.c:180]: We'll have to at some point if we want to support Suspend to RAM for I2C, but that alone is not sufficient. We'd also need to reinitialize the I2C registers after exiting S2RAM. So I'd say it's out of scope of this PR and can be done later in a dedicated S2RAM I2C PR.

**Review by gautierg-st — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by Oleh-Kravchenko** [comment]
Doesn't fix issue for STM32F1X.
This patch works for me https://github.com/zephyrproject-rtos/zephyr/issues/58929#issuecomment-1597208328

**Pull request merged**
